kaldb/src/test/java/com/slack/kaldb/com/slack/kaldb/logstore/search/KaldbDistributedQueryServiceTest.java

```diff
@@ -0,0 +1,132 @@
+package com.slack.kaldb.com.slack.kaldb.logstore.search;
+
+import static com.slack.kaldb.chunk.ChunkInfo.toSnapshotMetadata;
+import static com.slack.kaldb.chunk.ReadWriteChunk.LIVE_SNAPSHOT_PREFIX;
+import static com.slack.kaldb.chunk.ReadWriteChunk.toSearchMetadata;
+import static com.slack.kaldb.logstore.search.KaldbDistributedQueryService.getSnapshotsToSearch;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.spy;
+
+import brave.Tracing;
+import com.slack.kaldb.chunk.ChunkInfo;
+import com.slack.kaldb.chunk.ReadOnlyChunkImpl;
+import com.slack.kaldb.chunk.SearchContext;
+import com.slack.kaldb.metadata.search.SearchMetadata;
+import com.slack.kaldb.metadata.search.SearchMetadataStore;
+import com.slack.kaldb.metadata.snapshot.SnapshotMetadata;
+import com.slack.kaldb.metadata.snapshot.SnapshotMetadataStore;
+import com.slack.kaldb.metadata.zookeeper.MetadataStore;
+import com.slack.kaldb.metadata.zookeeper.ZookeeperMetadataStoreImpl;
+import com.slack.kaldb.proto.config.KaldbConfigs;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+import java.time.Instant;
+import java.util.Collection;
+import org.apache.curator.test.TestingServer;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class KaldbDistributedQueryServiceTest {
+
+  private SimpleMeterRegistry metricsRegistry;
+
+  private MetadataStore zkMetadataStore;
+  private SearchMetadataStore searchMetadataStore;
+  private SnapshotMetadataStore snapshotMetadataStore;
+  private TestingServer testZKServer;
+  private SearchContext indexerSearchContext;
+  private SearchContext cacheSearchContext;
+
+  @Before
+  public void setUp() throws Exception {
+    Tracing.newBuilder().build();
+
+    metricsRegistry = new SimpleMeterRegistry();
+    testZKServer = new TestingServer();
+
+    // Metadata store
+    KaldbConfigs.ZookeeperConfig zkConfig =
+        KaldbConfigs.ZookeeperConfig.newBuilder()
+            .setZkConnectString(testZKServer.getConnectString())
+            .setZkPathPrefix("indexerTest")
+            .setZkSessionTimeoutMs(1000)
+            .setZkConnectionTimeoutMs(1000)
+            .setSleepBetweenRetriesMs(1000)
+            .build();
+
+    zkMetadataStore = spy(ZookeeperMetadataStoreImpl.fromConfig(metricsRegistry, zkConfig));
+
+    snapshotMetadataStore = spy(new SnapshotMetadataStore(zkMetadataStore, true));
+    searchMetadataStore = spy(new SearchMetadataStore(zkMetadataStore, true));
+    indexerSearchContext = new SearchContext("host1", 10000);
+    cacheSearchContext = new SearchContext("host2", 20000);
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    snapshotMetadataStore.close();
+    searchMetadataStore.close();
+    zkMetadataStore.close();
+    metricsRegistry.close();
+    testZKServer.close();
+  }
+
+  @Test
+  public void testSnapshotsToSearch() throws Exception {
```

Comment:
Please add a few more unit tests to this code. 

* A case when data is live in indexer but no where else.
* A case when data is in indexer and cache.
* A cache when the same node has no live nodes, but multiple replicas in cache nodes. 
* A unit test where no nodes exist for a given query.
* A unit test to ensure if multiple services exist, we only pick the snapshots for a single service.
* A unit test to ensure that we are picking nodes randomly across multiple calls to the method.
* A unit test to document what happens to distributed query when this function fails.

Reply:
> A case when data is live in indexer but no where else.

Isn't that the very first test here?

> A case when data is in indexer and cache.

That's the second test

Reply:
let me break it down to multiple methods so it's easier for you to review

Reply:
done!